### PR TITLE
Fix setting `karpenter.sh/provisioner-name` to nodeSelector causes an error

### LINKED ISSUE #2320: Setting `karpenter.sh/provisioner-name` to nodeAffinity causes the pod to be PENDING
## Version

**Karpenter**: v0.14.0 and v0.15.0 (later than v0.13.2)

**Kubernetes**: v1.22.10 (`version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.10-eks-84b4fe6", GitCommit:"cc6a1b4915a99f49f5510ef0667f94b9ca832a8a", GitTreeState:"clean", BuildDate:"2022-06-09T18:24:04Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}`)

## Expected Behavior

As written in the docs (https://karpenter.sh/v0.15.0/faq/#if-multiple-provisioners-are-defined-which-will-my-pod-use), we use a feature to specify specific provisioner for the pods.

By writing manifest like below:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: karpenter.sh/provisioner-name
                operator: In
                values:
                - default-provisioner
      containers:
      - image: nginx
        imagePullPolicy: IfNotPresent
        name: nginx
        resources:
          requests:
            cpu: 4
      tolerations:
      - effect: NoSchedule
        key: provisioned-by-karpenter
        operator: Exists
```

(If the resource request is satisfied to already existing nodes, Karpenter skips scheduling, so that the resource requests should be large enough to trigger Karpenter to create new node.)

## Actual Behavior

The pods are worked as expected before v0.14.0, but the pods got `PENDING` status since v0.14.0 (and later).

The `karpenter` pod shows the DEBUG log:

```
controller 2022-08-19T04:45:17.747Z    DEBUG    controller.provisioning    Ignoring pod, label karpenter.sh/provisioner-name is restricted; specify a well known label: [karpenter.k8s.aws/instance-category karpenter.k8s.aws/instance-cpu karpenter.k8s.aws/instance-family karpenter.k8s.aws/instance-generation karpenter.k8s.aws/instance-gpu-count karpenter.k8s.aws/instance-gpu-manufacturer karpenter.k8s.aws/instance-gpu-memory karpenter.k8s.aws/instance-gpu-name karpenter.k8s.aws/instance-hypervisor karpenter.k8s.aws/instance-local-nvme karpenter.k8s.aws/instance-memory karpenter.k8s.aws/instance-pods karpenter.k8s.aws/instance-size karpenter.sh/capacity-type kubernetes.io/arch kubernetes.io/os node.kubernetes.io/instance-type topology.kubernetes.io/region topology.kubernetes.io/zone], ora custom label that does use a restricted domain: [k8s.io karpenter.k8s.aws karpenter.sh kubernetes.io]    {"commit": "3d87474", "pod": "XXX"}
```

## Steps to Reproduce the Problem

- Create a pod (or rollout, deployment, ...) with nodeAffinity that uses a key `karpenter.sh/provisioner-name`.

## Resource Specs and Logs

I think just setting nodeAffinity is the cause, but if needed, please let me know!

### PR DESCRIPTION
**Pull request by kyontan**
Fixes #2320

**Description**

We got an error when we upgraded Karpenter to v0.14.0 from v0.13.2. (Also v0.15.0 shows the same error)

```
controller 2022-08-19T04:45:17.747Z    DEBUG    controller.provisioning    Ignoring pod, label karpenter.sh/provisioner-name is restricted; specify a well known label: [karpenter.k8s.aws/instance-category karpenter.k8s.aws/instance-cpu karpenter.k8s.aws/instance-family karpenter.k8s.aws/instance-generation karpenter.k8s.aws/instance-gpu-count karpenter.k8s.aws/instance-gpu-manufacturer karpenter.k8s.aws/instance-gpu-memory karpenter.k8s.aws/instance-gpu-name karpenter.k8s.aws/instance-hypervisor karpenter.k8s.aws/instance-local-nvme karpenter.k8s.aws/instance-memory karpenter.k8s.aws/instance-pods karpenter.k8s.aws/instance-size karpenter.sh/capacity-type kubernetes.io/arch kubernetes.io/os node.kubernetes.io/instance-type topology.kubernetes.io/region topology.kubernetes.io/zone], ora custom label that does use a restricted domain: [k8s.io karpenter.k8s.aws karpenter.sh kubernetes.io]    {"commit": "3d87474", "pod": "XXX"}
```

We uses a feature to use specific provisioner by setting `karpenter.sh/provisioner-name` label to nodeAffinity of the pods.
The feature is described in FAQs: https://karpenter.sh/v0.15.0/faq/#if-multiple-provisioners-are-defined-which-will-my-pod-use .

- Fix: allow setting `karpenter.sh/provisioner-name` to nodeAffinity
- Fix typo in error message

**How was this change tested?**

* Tested and confirmed that the problem is fixed on our EKS cluster

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
fix: Fix setting `karpenter.sh/provisioner-name` to nodeAffinity blocks node creation
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/provisioning/v1alpha5/labels.go [hunk 1]
```diff
@@ -89,7 +90,7 @@ func IsRestrictedLabel(key string) error {
 		return nil
 	}
 	if IsRestrictedNodeLabel(key) {
-		return fmt.Errorf("label %s is restricted; specify a well known label: %v, or a custom label that does use a restricted domain: %v", key, WellKnownLabels.List(), RestrictedLabelDomains.List())
+		return fmt.Errorf("label %s is restricted; specify a well known label: %v, or a custom label that does not use a restricted domain: %v", key, WellKnownLabels.List(), RestrictedLabelDomains.List())
```
Comment (kyontan) [inline pkg/apis/provisioning/v1alpha5/labels.go:93]: This is just a minor debug message fix.

**Comment by ellistarn** [comment]
This is a great catch! Do you mind updating this test case here: https://github.com/aws/karpenter/blob/main/test/suites/integration/scheduling_test.go#L62

I think we also probably need a test (and maybe some logic) that you can't create a Provisioner with this label.

**Comment by kyontan** [comment]
Hi @ellistarn, thanks for your suggestion!

I tried to a

```
# kubectl patch provisioner default-provisioner --type='json' -p='[{"op": "add", "path": "/spec/labels", "value": {"foo": "bar" } }]'
provisioner.karpenter.sh/default-provisioner patched

# kubectl patch --type merge provisioner default-provisioner -p='{"spec": {"labels": {"karpenter.sh/provisioner-name": "foo" }}}'
Error from server (BadRequest): admission webhook "validation.webhook.provisioners.karpenter.sh" denied the request: validation failed: invalid key name "karpenter.sh/provisioner-name": spec.labels
label karpenter.sh/provisioner-name is restricted; specify a well known label: [karpenter.k8s.aws/instance-category karpenter.k8s.aws/instance-cpu karpenter.k8s.aws/instance-family karpenter.k8s.aws/instance-generation karpenter.k8s.aws/instance-gpu-count karpenter.k8s.aws/instance-gpu-manufacturer karpenter.k8s.aws/instance-gpu-memory karpenter.k8s.aws/instance-gpu-name karpenter.k8s.aws/instance-hypervisor karpenter.k8s.aws/instance-local-nvme karpenter.k8s.aws/instance-memory karpenter.k8s.aws/instance-pods karpenter.k8s.aws/instance-size karpenter.sh/capacity-type kubernetes.io/arch kubernetes.io/os node.kubernetes.io/instance-type topology.kubernetes.io/region topology.kubernetes.io/zone], or a custom label that does use a restricted domain: [k8s.io karpenter.k8s.aws karpenter.sh kubernetes.io]
```

File: pkg/cloudprovider/aws/apis/v1alpha1/register.go [hunk 2]
```diff
@@ -81,7 +81,7 @@ var (
 func init() {
 	Scheme.AddKnownTypes(schema.GroupVersion{Group: v1alpha5.ExtensionsGroup, Version: "v1alpha1"}, &AWS{})
 	v1alpha5.RestrictedLabelDomains = v1alpha5.RestrictedLabelDomains.Insert(RestrictedLabelDomains...)
-	v1alpha5.WellKnownLabels = v1alpha5.WellKnownLabels.Insert(
```
Comment (kyontan) [inline pkg/cloudprovider/aws/apis/v1alpha1/register.go:84]: NOTE: `sets.Insert()` updates the receiver itself, so no need to assign.
https://github.com/kubernetes/apimachinery/blob/f916759cb6b8547418dc7708876ecab5c1961448/pkg/util/sets/string.go#L48-L54

File: pkg/apis/provisioning/v1alpha5/labels.go [hunk 3]
```diff
@@ -53,10 +53,24 @@ var (
 		v1.LabelNamespaceSuffixNode,
 	)
 
-	// WellKnownLabels are labels that belong to the RestrictedLabelDomains but allowed.
-	// Karpenter is aware of these labels, and they can be used to further narrow down
-	// the range of the corresponding values by either provisioner or pods.
-	WellKnownLabels = sets.NewString(
+	// WellKnownLabelsForProvisioner are labels that belong to the RestrictedLabelDomains but
+	// allowed in provisioner. Karpenter is aware of these labels.
+	WellKnownLabelsForProvisioner = sets.NewString(
```
Comment (ellistarn) [inline pkg/apis/provisioning/v1alpha5/labels.go:58]: Sorry if this is a bit confusing, but WellKnownLabels for provisioner is identical to WellKnownLabels except for `ProvisionerNameLabelKey` However, the `WellKnownLabels` public variable is actually mutated by cloud providers to add provider specific labels, so we need to be a little careful about duplicating the slices. 

Instead, I think we can fix this by adding another validation (https://github.com/aws/karpenter/blob/main/pkg/apis/provisioning/v1alpha5/provisioner_validation.go#L91) to check that the provisioner name label isn't set. All other labels between pods/provisioners will be identical, so we can just rely on a single WellKnownLabels variable.

Comment (ellistarn) [reply pkg/apis/provisioning/v1alpha5/labels.go:58]: You should be able to test this by creating a provisioner with a provider specific label like `karpenter.k8s.aws/instance-family`.

**Comment by ellistarn** [comment]
I wrote up a quick PR here that does what you're suggesting, plus my recommendations: https://github.com/aws/karpenter/pull/2328/files. Feel free to just rebase it and submit, so you get the commit credit for finding the bug :).

File: pkg/apis/provisioning/v1alpha5/labels.go (hunk 3 above)
Comment (kyontan) [reply pkg/apis/provisioning/v1alpha5/labels.go]: @ellistarn Ah, thank you for following up! I agree with you 🙏 
> However, the WellKnownLabels public variable is actually mutated by cloud providers to add provider specific labels, so we need to be a little careful about duplicating the slices.

Also, many thanks for your helps in https://github.com/aws/karpenter/pull/2328, I fixed this PR based on the pull request (and marked the commit with co-authored with you.)

**Comment by ellistarn** [comment]
Looks like @bwagner5 merged my PR. Sorry for the churn, and thanks so much for identifying this bug!

**Comment by kyontan** [comment]
@ellistarn No worries! Thanks for fixing the issue!

**Pull request closed without merging**